### DISCUSSION TITLE
Testing RelationManagers

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#2189] [Help]
For regular resources I usually employ two tests to ensure that my resources are rendering and pulling in the data needed. For example:

```
/**
 * @test
 */
public function it_will_list_the_camp_events()
{
    $campevent = CampEvent::factory()->create();
    $this->assertEquals(CampEvent::withoutGlobalScopes(), CampEventResource::getEloquentQuery());
    $this->assertEquals($campevent->id, CampEventResource::getEloquentQuery()->first()->id);
}

/**
 * @test
 */
public function it_will_display_the_correct_data_when_listing_camp_events()
{
    $campevent = CampEvent::factory()->create();

    Livewire::test(ListCampEvents::class)
        ->assertSee($campevent->date->format('Y-m-d'))
        ->assertSee($campevent->places_available)
        ->assertSee($campevent->capacity - $campevent->places_available)
        ->assertSee($campevent->capacity);
}
```
For relation managers, the second item is fine to replicate but the first is tricky as `RelationManager` doesn't have a static function for `getEloquentQuery` plus I'd need to pass in the `ownerRecord` somehow and I couldn't see anything equivalent. I know the test may seem like overkill but I do it for completeness and it's useful to run if the relationship changes for some reason. Also good to check it here than rely on `assertSee` which I find a bit "flakey". Is there a method to retrieve the Eloquent query used from a RelationManager for use within a test like this?

**Comment by danharrin** [comment]
Ryan is working on some testing helpers for tables (and relation managers) right now :)

**Reply by titantwentyone** [reply]
Brilliant. @ryangjchandler if you need some help with test driving this out, give me a shout.

**Comment by danharrin** [comment]
@titantwentyone I have some news...

https://filamentphp.com/docs/2.x/tables/testing

**Reply by titantwentyone** [reply]
Oh wow! This looks awesome. Thank you!

**Reply by danharrin** [reply]
Let me know if you're missing anything, we can easily add more stuff.

**Reply by cubizinfotech** [reply]
Hi @danharrin 

I have facing some issues can you resolve it and share solution.

it('can hard delete device key relation manager', function () {
    $device = Device::factory()->has(DeviceKey::factory()->count(1))->create();
    $deviceKey = $device->deviceKeys->first();
    $id = $deviceKey->id;

    // Soft delete first
    $deviceKey->delete();
    expect(DeviceKey::withTrashed()->find($id))->not->toBeNull(); // Ensure soft deletion worked

    Livewire::test(DeviceKeysRelationManager::class, [
        'ownerRecord' => $device,
        'pageClass' => EditDevice::class,
    ])
    // ->assertTableActionExists('forceDelete')
    ->callTableAction(ForceDeleteAction::class, $deviceKey); // Correctly call the action with the record ID
    // ->assertHasNoTableActionErrors();

    $this->assertModelMissing($deviceKey);

    // Now, the record should be completely gone
    // expect(DeviceKey::withTrashed()->find($id))->toBeNull();
});

**Error:**

  FAILED  Tests\Feature\Filament\DeviceResourceTest > it can hard delete device key relation manager
  Failed asserting that a row in the table [device_keys] does not match the attributes {
    "id": 1
}.

Found similar results: [
    {
        "id": 1
    }
].

  at tests\Feature\Filament\DeviceResourceTest.php:294
    290▕     // ->assertTableActionExists('forceDelete')
    291▕     ->callTableAction(ForceDeleteAction::class, $deviceKey); // Correctly call the action with the record ID
    292▕     // ->assertHasNoTableActionErrors();
    293▕
  ➜ 294▕     $this->assertModelMissing($deviceKey);
    295▕
    296▕     // Now, the record should be completely gone
    297▕     // expect(DeviceKey::withTrashed()->find($id))->toBeNull();
    298▕ });


  Tests:    1 failed (4 assertions)
  Duration: 3.39s

**Comment by cubizinfotech** [comment]
https://github.com/filamentphp/filament/discussions/2189#discussioncomment-12613312
